les, les/lespay: implement new server pool

zsfelfoldi opened the pull request:
This PR reimplements the light client server pool. It is also a first step to move certain logic into a new `lespay` package. This package will contain the implementation of the `lespay` token sale functions, the token buying and selling logic and other components related to peer selection/prioritization and service quality evaluation. Over the long term this package will be reusable for incentivizing future protocols.
The new server pool is only partially implemented in this PR, this is why some components (like `NodeStateMachine`) might seem a bit overkill for the task. Subsequent PRs will implement the pre-negotiation filtering and token buying/paid server selecting logic (based on the `les/4` token sale mechanism). It provides an `enode.Iterator` output and also requires an iterator input for newly discovered nodes. Currently only the new DNS discovery is used as an input because `discv5` will be replaced very soon. Later results from DNS and the new `discv5` mixed together will be used as the input.

This document describes the funtion of the new components:
https://gist.github.com/zsfelfoldi/3c7ace895234b7b345ab4f71dab102d4

rjl493456442 commented on les/lespay/utils/nodestate.go:
```diff
@@ -0,0 +1,398 @@
+		ns.lock.Lock()
+		if t.mask != 0 {
+			fmt.Println("timeout", id, "mask", ns.stateToString(t.mask))
+			ns.updateState(id, 0, t.mask, 0)
```
cb = updateState(id, 0, t.mask, 0)

rjl493456442 commented:
Can we move `les/lespay/utils` to `les/utils`? Since these common components can be used in all les modules.

rjl493456442 commented:
The idea of nodeStateMachine is awesome, but the biggest issue is: it's very fragile. Whenever the "Tag" order we initialize in the state machine changed, the persisted value is useless(and even worse)

As far as I see, there are already some mistakes made. Like the "Tag": (`iterSelected`), (`iterReturned`)
never be assigned(actually it's `selected`, `returned`)

rjl493456442 commented on les/utils/nodestate.go:
```diff
@@ -0,0 +1,419 @@
+		ns.lock.Lock()
+		if t.mask != 0 {
+			fmt.Println("timeout", id, "mask", ns.stateToString(t.mask))
+			cb = ns.updateState(id, 0, t.mask, 0)
```
Just a question: So if some state are set(and timeout is specified), after the given time duration, the state will be reset and relevant callbacks are called?

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,283 @@
-	} else {
-		pool.newSelected--
+	rts := s.vt.RtStats()
+	rts.Add(time.Second*2, 10, s.vt.StatsExpFactor())
```
Any particular reason to add this "fake" statistic?

zsfelfoldi replied:
It is an easy way to initialize with some conservative values when the database is new and we have no statistics. As soon as we have a considerable amount of real stats this small value won't matter.

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,283 @@
-	if err := s.Decode(&entry); err != nil {
-		return err
+	if t := rts.Timeout(0.5) * 2; t > timeout {
+		timeout = t
```
I can't understand the logic here. 10 percent failure ratio seems conservative, 50 percent failure ratio seems aggressive. 

The logic here:
* If the conservative `to` is even smaller than minimal one, use minimal; otherwise record conservative `to`.
* If the aggressive `to` is larger than `conservative` to, then use aggressive one.

zsfelfoldi replied:
`rts.Timeout(0.5)` is multiplied by 2, this is why it can be larger than the conservative timeout (otherwise this would not be possible). So we either use the 10% failure rate timeout or twice the median response time, whichever is more conservative. We can't really predict what kind of distribution we are going to have (might also depend on the connection type) so I am mostly making a guess at this point but I think twice the median is usually not too high and hopefully it can produce even less than 10% failure rate if the response times are predictable and the distribution is narrow. We do check it against the 10% failure threshold though so we are not going to have anything worse than that over the long term even if there are lots of slow responses and we have a "long tail" distribution.

rjl493456442 commented:
Btw, now the les won't maintain the nodedb right? We totally rely on the underlying network system to provide us some "cached" nodes?

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,283 @@
+	sfiNodeHistory = utils.NewNodeStateField("nodeHistory", reflect.TypeOf(&nodeHistory{}), []*utils.NodeStateFlag{sfDiscovered, sfHasValue},
+		func(field interface{}) ([]byte, error) {
+			if n, ok := field.(*nodeHistory); ok {
+				enc, err := rlp.EncodeToBytes(&n.dialCost)
```
Why pass the integer pointer? Maybe `rlp.EncodeToBytes(n.dialCost)`?

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,283 @@
+	s.stConnected = s.ns.MustRegisterState(sfConnected)
+	s.stRedialWait = s.ns.MustRegisterState(sfRedialWait)
+	s.stAlwaysConnect = s.ns.MustRegisterState(sfAlwaysConnect)
+	s.ns.MustRegisterState(sfSelected)
```
Can we also save the `selected flag indicator` here? Then we can pass it to different iterators and they don't need to obtain the indicator anymore

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,283 @@
-	}
-	return nodes
+		n.lock.Lock()
+		n.dialCost.Add(dialCost, s.vt.StatsExpirer().LogOffset(s.clock.Now()))
```
let's make it a member function of `nodeHistory`. So that we don't need to do `LOCK` `UNLOCK` opts.

zsfelfoldi replied:
Now `nodeHistory` is stored in its field as a value instead of a reference so we don't need locking any more.

zsfelfoldi commented:
@fjl maybe we are misunderstanding each other here but I think the issues you are bringing up about the node state set are already addressed in this code. There is a statically constructed `NodeStateSetup`. The version numbering also exists, though it is automatically handled inside `NodeStateMachine` (see `mappings` and `currentMapping` index). Saved bit masks and field orders are automatically converted if the `NodeStateSetup` is different from the one used when saving. Bit masks are always generated runtime after creating `NodeStateMachine` (by `StateMask` or `StatesMask`) so unless you are explicitly converting `uint64` to `NodeStateBitMask` (which you should not do) there is no risk of messing up the bit mapping. I think it is nice to use bit masks everywhere during runtime operation. Actually I find it safer to convert everything to bit masks at the beginning because that ensures that anything being messed up in the state logic will generate an error right at the start and not when some rare case happens.
I agree with your other comments and will change them.

fjl commented:
Sorry, I didn't see that you already did this change in the diff. Still think it would be better to assign the masks while the `Setup` is constructed instead of doing that part in the machine. 

I'm also going to be a little bit insistent about using flags over bitmasks in the API because it allows you to remove the bitmask struct fields in server pool. There is really no reason to have both flags and mask as separate things if you assign the masks statically, and it will be much less effort to change the states later because the struct doesn't have to change when there is a new state.

I agree the bitmask stuff is a nice API sometimes, and I'm not saying the bitmask should go away. It's just that the flag is a struct and not just a number, so you can put more stuff in there including the bitmask.

fjl commented:
I'm not so much afraid about messing up the mapping by converting from `uint64` to `NodeStateBitMask`. I'm more afraid that someone might use multiple state machines in the same general area and then use a state mask with the wrong machine. It also really bugs me that every state has two ways of referring to it :).

zsfelfoldi commented:
The `NodeStateSetup` does define the bit order so it is not assigned runtime any more. I do agree that it was a bad design originally but I think the current one works as you want it to. I can be convinced about the API too :) So just to get on the same page before I start making code changes: what you suggest is that we do not export the bit mask type at all? It might work because most of the convenience the bit masks are giving me is inside the state machine logic. So I can still use an unexported `nodeStateBitMask` internally and use `[]*nodestate.Flag` instead of bit masks in the API? (also `*nodestate.Field` instead of field index) If this is what you suggest then it does make sense to me too.

fjl commented:
The change I'm suggesting is mostly that `NodeStateMachine.StateFlag` and `StatesFlag` should not exist the way they do now.

Using `[]*nodestate.Flag` in the API will make the call sites uglier because you'll
need to make slices of states all the time. Maybe the API should just take `*nodestate.Flag`
and then we add an operation that combines multiple flags into a single one? A bit like `StatesMask`, but more 'static'.

```go
func Mask(states ...*Flag) *Flag
```

With that, if you want to remove two states from the node, it's

```go
ns.SetState(0, nodestate.Mask(sfDialed, sfConnected), 0)
```

vs.

```go
ns.SetState(0, s.stDialed|s.stConnected, 0)
```

zsfelfoldi commented:
We still have to come up with a nice way to handle subscription callbacks though because they receive the old and new states (masked with the subscription mask). Inside my iterators it is very easy to use bit masks for required and disabled states. I do not care how many flags are in those masks, just use binary operations to check whether we moved into or out of the allowed set.
Maybe we could create an abstract "flag set" interface that can be satisfied either by a single `*nodestate.Flag` or by a `struct { mask nodeStateBitMask, ns *NodeStateMachine }` structure. I could allow binary operations but would also be able to check whether it is created by the same state machine. Any thoughts on that?

zsfelfoldi commented:
I see you were suggesting something similar while I was typing :)

fjl commented:
I think the mask should still be used for the subscribe callback, since multiple
states can transition at the same time.

```go
ns.SubscribeState(s.stDialed|s.stConnected, func(n *enode.Node, oldState, newState utils.NodeStateBitMask) {
    if oldState == s.stDialed && newState == 0 {
```

This can still look smooth if you add a helper method to test the mask with a flag:

```go
ns.SubscribeState(nodestate.Mask(sfDialed, sfConnected), func(n *enode.Node, oldState, newState nodestate.Mask) {
    if oldState.Is(sfDialed) && newState == 0 {
```

i.e. 

```go
func (m Mask) Is(f *Flag) bool {
    return m & f.mask != 0
}
```

fjl commented:
The alternative is to just leave the mask like it is for the subscriptions. We can still use bitmasks, we just lose a tiny bit of type safety.

fjl commented:
I wouldn't use an interface for this. It will mess up the implementation a lot.

zsfelfoldi commented:
Right, using them in the callback is probably fine because there it is very obvious which `NodeStateMachine` they belong to.

fjl commented:
One more thing: the implementation becomes pretty trivial if you do it like this:

```go
type Flag {
    mask Mask
    setup Setup
}
```

With this type definition, Flag is just two machine words and you probably don't need to pass it around as a pointer anymore. Then `Mask` is just

```go
func Mask(flags ...Flag) Flag {
    var r Flag
    for f, _ := range flags {
        r.mask |= f.mask
        r.setup = f.setup // and maybe check setup
    }
}
```

If you want `String`, you'd need to store the flag names in the Setup and implement it through that.

fjl commented:
I've pushed some small changes now:

- With `--syncmode=light` geth now uses the les-specific DNS list. This makes pre-filtering the iterator output unnecessary.
- The new iterators now use `sync.Cond` instead of the `wakeup` channel. This doesn't work 100% for the WRS iterator because it looks like the WRS chooses `nil` sometimes even though it should be non-empty.

fjl commented:
I also think we should move the construction of the state machine into serverpool.

fjl commented:
Not sure if I misunderstand the weighted random selection code, but the test for WrsIterator occasionally fails because `Choose` seems to just remove the one item that's in the selection tree:

Before `Choose`:

```
(*utils.WeightedRandomSelect)(0xc00027e4a0)({
 root: (*utils.wrsNode)(0xc0002941c0)({
  items: ([8]interface {}) (len=8 cap=8) {
   (interface {}) <nil>,
   (enode.ID) (len=32 cap=32) 2a02000000000000000000000000000000000000000000000000000000000000,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>
  },
  weights: ([8]uint64) (len=8 cap=8) {
   (uint64) 0,
   (uint64) 1,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0
  },
  sumWeight: (uint64) 1,
  level: (int) 0,
  itemCnt: (int) 1,
  maxItems: (int) 8
 }),
 idx: (map[utils.WrsItem]int) (len=1) {
  (enode.ID) (len=32 cap=32) 2a02000000000000000000000000000000000000000000000000000000000000: (int) 1
 },
 wfn: (utils.WeightFn) 0x42344b0
})
```

`Choose` returns `nil`, and the selection tree is updated to

```
(*utils.WeightedRandomSelect)(0xc00027e4a0)({
 root: (*utils.wrsNode)(0xc0002941c0)({
  items: ([8]interface {}) (len=8 cap=8) {
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>,
   (interface {}) <nil>
  },
  weights: ([8]uint64) (len=8 cap=8) {
   (uint64) 0,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0,
   (uint64) 0
  },
  sumWeight: (uint64) 0,
  level: (int) 0,
  itemCnt: (int) 0,
  maxItems: (int) 8
 }),
 idx: (map[utils.WrsItem]int) {
 },
 wfn: (utils.WeightFn) 0x42344b0
})
```

zsfelfoldi commented:
The wrs indeed does that. The logic here is that the weight of the item in the tree is an upper estimate, the actual weight can continuously decrease over time, without frequently recalculating and updating it in the tree. Whenever an item is selected based on the estimate, the current weight is recalulated using the weight function. So it is possible that a wrs tree first seems to have some items with a positive weight but after selecting and recalculating them they have zero weight and we cannot select anything in the end. In this case you should just keep waiting for an actual selectable item.
I am not sure yet why the test fails though; the code seems to correctly go back to waiting if the selection fails. I will try to debug it.

fjl commented:
> I am not sure yet why the test fails though; the code seems to correctly go back to waiting if the selection fails. I will try to debug it.

This is because I pushed a fix for the issue.

rjl493456442 commented on p2p/nodestate/nodestate.go:
```diff
@@ -0,0 +1,966 @@
+	for _, t := range node.timeouts {
+		storedState &= ^t.mask
+	}
+	if storedState == 0 {
```
I think in this case, these timeouts are applied in advance. So we should get rid of them.
Otherwise it might be buggy this the following example:

- Addtimeout(node, flags)
- Persist(node)
- UpdateState(node, flags')
- Timeout, the re-added flags are removed

zsfelfoldi replied:
`Persist` does not change the node state so the flags with the active timeouts are not actually removed and therefore no inconsistency is created here. These flags will stay active until the originally planned timeout. Here we are just calculating the state of the stored version which only includes persistent flags with no timeout.

zsfelfoldi commented on les/serverpool.go:
```diff
@@ -17,904 +17,364 @@
+
+	sfDisableSelection = nodestate.MergeFlags(sfSelected, sfDialed, sfConnected, sfRedialWait)
+
+	errInvalidField = errors.New("invalid field type")
```
done

zsfelfoldi commented on p2p/nodestate/nodestate.go:
```diff
@@ -0,0 +1,966 @@
+
+// stateMask checks whether the set of flags belongs to the same setup and returns its internal bit mask
+func (ns *NodeStateMachine) stateMask(flags Flags) bitMask {
+	if flags.setup != ns.setup && flags.mask != 0 {
```
This condition allows using `nodestate.Flags{}` as an empty set of flags anywhere. It makes sense because in other cases the reference ensures the correct interpretation of the flag bits but an empty set is pretty universal so there is no need to create it from the specific `nodestate.Setup`.

rjl493456442 replied:
make sense

zsfelfoldi commented on p2p/nodestate/nodestate.go:
```diff
@@ -0,0 +1,966 @@
+		return
+	}
+	// convertMask converts a old format state/mask to the latest version.
+	convertMask := func(schemeID int, scheme []string, state bitMask) (bitMask, bool) {
```
done

zsfelfoldi commented on p2p/nodestate/nodestate.go:
```diff
@@ -0,0 +1,966 @@
+func (ns *NodeStateMachine) Persist(n *enode.Node) error {
+	ns.lock.Lock()
+	defer ns.lock.Unlock()
+
```
done

zsfelfoldi commented on les/lespay/client/wrsiterator.go:
```diff
@@ -0,0 +1,122 @@
+
+		if ns {
+			w.wrs.Update(n.ID())
+			if w.wakeup != nil && !w.wrs.IsEmpty() {
```
done

zsfelfoldi commented on les/utils/nodestate.go:
```diff
@@ -0,0 +1,909 @@
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
```
done

zsfelfoldi commented on les/lespay/client/queueiterator.go:
```diff
@@ -0,0 +1,120 @@
+				if qi.wakeup != nil {
+					close(qi.wakeup)
+					qi.wakeup = nil
+				}
```
done

zsfelfoldi commented on les/lespay/client/queueiterator.go:
```diff
@@ -0,0 +1,131 @@
+			qi.nextID = qi.queue[0]
+			e := qi.ns.GetField(qi.nextID, qi.enrFieldID)
+			var ok bool
+			if qi.nextENR, ok = e.(*enr.Record); ok {
```
Now the ENRs are handled by `NodeStateMachine` because this logic was indeed ugly.

zsfelfoldi commented on les/lespay/client/wrsiterator.go:
```diff
@@ -0,0 +1,131 @@
+			w.nextID = n.(enode.ID)
+			e := w.ns.GetField(w.nextID, w.enrFieldID)
+			var ok bool
+			if w.nextENR, ok = e.(*enr.Record); ok {
```
Now the ENRs are handled by `NodeStateMachine` because this logic was indeed ugly.

zsfelfoldi commented:
@fjl @rjl493456442 I think I addressed all issues. I also added the PreNegFilter because I wanted to see and test the whole logic together. I think it is also ready for review, I added docs and a test. It is just 200 lines so maybe we can merge the whole thing in a single step but if you prefer merging without the filter first then I can easily remove it from this PR (it is optional anyway).

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,354 @@
-		select {
-		case <-pool.closeCh:
-		case <-time.After(dialTimeout):
+		for {
```
I think we can remove this go-routine.

We don't need to persist the clock. Also we can update the metrics whenever we update the `timeout` and `timeweight`. otherwise it's weird to update metrics every 5 minutes.

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,354 @@
-			pool.knownSelect.Update((*knownEntry)(e))
+	iter := enode.Iterator(s.mixer)
+	if query != nil {
+		var testClock *mclock.Simulated
```
Please check my PR against your repo

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,354 @@
+}
+
+// updateNode calculates the selection weight and the proposed redial wait time of the given node
+func (s *serverPool) updateNode(node *enode.Node, calculateSessionValue, redialWait bool, addDialCost int64) {
```
I would recommend to split this big function into several parts.

For example `recalSessionValue`, `recalRedialInterval`, etc. It's not easy to determine the functionality of this function, with lots of paramters.

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,394 @@
+		diff.SubStats(&connStats)
+		sessionValue = diff.Value(timeWeights, expFactor)
+		sessionValueMeter.Mark(int64(sessionValue))
+		s.ns.SetField(node, sfiConnectedStats, nil)
```
Why set it to nil? Maybe `s.ns.SetField(node, sfiConnectedStats, currentStats)` ?

zsfelfoldi replied:
The purpose of this flag is not to act as a "cache" of `RtStats` but to calculate the change between connection and disconnection. It serves no purpose when a node is offline, we are going to set it during next connection anyway so I thought it is nicer to just remove it explicitly and free the allocated memory.

rjl493456442 replied:
I still don't understand. The `sfiConnectedStats` can be used to calculate the session value, but why set it to nil?

If we call `serviceValue` first Time, we can get a meaningful session value. But the next time, it's always 0.

zsfelfoldi replied:
Oh, I see, this was a mistake. I forgot that `unregisterPeer` already resets this flag (which is a much better place to do it). Resetting it in `serviceValue` caused no harm because it is only called once but it is unnecessary. I'll remove this line.

zsfelfoldi replied:
done

rjl493456442 commented on les/serverpool.go:
```diff
@@ -17,904 +17,393 @@
+	// because the pre-negotiation query did not return a positive result. In this case
+	// the ratio has no meaning anyway and waitFactor is always raised, though in smaller
+	// steps because queries are cheaper and therefore we can allow more failed attempts.
+	a := totalValue * dialCost
```
Can we set a hard limitation for redial interval. In theory if the peer doesn't behave as the expected, we can refuse to dial it. However, my concern is if we have some bugs which always try to disconnect the peer, then the interval will be very high.

zsfelfoldi replied:
Though it is increased exponentially with each dial attempt, as a function of elapsed time it increases linearly even if it is never accessible so I don't think there is a reason to hard limit it. Instead maybe we can discard it if it hasn't been seen in the last few months. I'll think about that.

zsfelfoldi replied:
Now I realized you are right and it is indeed possible that `redialWait` goes very high quickly because the node can be selected by `FairMix` directly from the discovery iterator. So it is dialed often and still the waiting time is raised every time.

zsfelfoldi replied:
fixed

zsfelfoldi commented on eth/discovery.go:
```diff
@@ -25,17 +25,17 @@ import (
+// EthEntry is the "eth" ENR entry which advertises eth protocol
 // on the discovery network.
-type ethEntry struct {
+type EthEntry struct {
```
No idea, probably there was a reason originally but there is none any more, so I reverted it.

zsfelfoldi commented on les/client.go:
```diff
@@ -159,6 +154,20 @@ func New(ctx *node.ServiceContext, config *eth.Config) (*LightEthereum, error) {
 	}
 	leth.ApiBackend.gpo = gasprice.NewOracle(leth.ApiBackend, gpoParams)
 
+	dnsdisc, err := leth.setupDiscovery(&ctx.Config.P2P)
```
done

rjl493456442 commented on les/serverpool.go:
```diff
@@ -239,9 +242,9 @@ func (s *serverPool) start() {
+		if n, ok := s.ns.GetField(node, sfiNodeHistory).(nodeHistory); ok && n.redialWaitEnd > unixTime {
+			wait := n.redialWaitEnd - unixTime
+			lastWait := n.redialWaitEnd - n.redialWaitStart
 			if wait > lastWait {
```
Can we simplify the logic as `wait := n.redialWaitEnd - unixTime`?

zsfelfoldi replied:
The rest of the logic is just a protection against setting crazy high timeouts if the system clock was adjusted. In the normal case (if `unixTime > n.redialWaitStart`) it can't hurt anything so I think it should stay.

rjl493456442 commented on les/serverpool.go:
```diff
@@ -96,6 +96,7 @@ var (
 	sfQueried          = serverPoolSetup.NewFlag("queried")
 	sfCanDial          = serverPoolSetup.NewFlag("canDial")
 	sfDialed           = serverPoolSetup.NewFlag("dialed")
+	sfDialTimeout      = serverPoolSetup.NewFlag("dialTimeout")
```
Ah, `sfDialTimeout` seems not a good name. Let's rename it to `sfWaitDialTimeout`

zsfelfoldi replied:
done

rjl493456442 commented on les/serverpool.go:
```diff
@@ -96,6 +96,7 @@ var (
 	sfQueried          = serverPoolSetup.NewFlag("queried")
 	sfCanDial          = serverPoolSetup.NewFlag("canDial")
 	sfDialed           = serverPoolSetup.NewFlag("dialed")
```
Now the `sfDialed` seems not a good name, let's rename is to `sfDialing`

zsfelfoldi replied:
done

rjl493456442 commented on les/serverpool.go:
```diff
@@ -96,6 +96,7 @@ var (
 	sfQueried          = serverPoolSetup.NewFlag("queried")
```
Now the `sfQuery` seems not a good name, let's rename it to `sfCanQuery`

zsfelfoldi replied:
When this flag is set the query is always and instantly started so while the flag is active it does not represent a state where a query might be started, rather one where it has been started. So I'd leave this one as is.
